docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,303 @@
+circumstance. Consider the statements:
+
+```sql
+  CREATE TABLE test (key INT PRIMARY KEY, value INT);
+  CREATE INDEX foo ON test (value);
+  INSERT INTO test VALUES (1, 2);
+```
+
+Depending on when the gossip of the schema change is received, the
+`INSERT` statement may either see no cached descriptor for the `test`
+table (which is safe because it will fall back to reading from the KV
+store), the descriptor as it looks after the `CREATE TABLE` statement
+or the descriptor as it looks after the `CREATE INDEX` statement. If
+the `INSERT` statement does not see the `CREATE INDEX` statement then
+it will merrily insert the values into `test` without creating the
+entry for the `foo` index, thus leaving the table data in an
+inconsistent state.
+
+It is easy to show similar problematic sequences of operations for
+`DELETE`, `UPDATE` and `SELECT`. Fortunately, a solution exists in the
+literature: break up the schema change into discrete steps for which
+any two consecutive versions of the schema are safe for concurrent
+use. For example, to add an index to a table we would perform the
+following steps:
+
+* Add the index to the table descriptor but mark it is as
+  write-only. Insert, update and delete operations would add entries
+  for the index, but read operations would not use the index.
+* Wait for the table descriptor change to propagate to all nodes in
+  the cluster.
+* Backfill the index.
+* Mark the index in the table descriptor as read-write.
+* Wait for the table descriptor change to propagate to all nodes in
+  the cluster.
+
+This RFC is focused on how to wait for the table descriptor change
+propagate to all nodes in the cluster. Additionally, we need
+mechanisms to ensure that multiple updates to a table descriptor are
+not performed concurrently and that write operations never commit once
+a table descriptor is too old (i.e. not one of the two most recent
+versions).
+
+At a high-level, we want to provide read and write leases for table
+descriptors. When a schema change is to be performed, the operation
+first needs to grab a write lease for the descriptor. The write lease
+ensures that no concurrent schema change will be performed and that
+there is only one active version of the descriptor in the cluster.
+Note that these are essentially the same constraint: there are only
+two active versions of a descriptor during a schema change.
+
+Before using a table descriptor for a DML operation (i.e. `SELECT`,
+`INSERT`, `UPDATE`, `DELETE`, etc), the operation needs to obtain a
+read lease for the descriptor. The lease will be guaranteed valid for
+a significant period of time (on the order of minutes) and the lease
+period can be extended as long as a scheme change is not
+performed. When the operation completes it will release the read
+lease.
+
+# Detailed design
+
+Leases will be managed and dispensed by a centralized lease
+service. Initially there will be a single lease service master, but a
```

Comment:
This is a major architectural change which violates the nodes are symmetric principle. Perhaps an alternative is to split this up at the granularity of a table, with a goroutine being spawned  to guide a schema change to a specific table. I also like the idea of all state changes and leases going through the KV store, so that they can be applied as transactions. I don' think we need a  service here.